File: apps/ios/Pebbles/Features/Record/RecordStepScaffold.swift
import SwiftUI

/// The single action a step may offer beneath its content.
///
/// Top-level rather than nested inside the generic scaffold: `RecordFlowView`
/// builds these in a `switch` that never names a `Content` type, and
/// `RecordStepScaffold<AnyView>.Action` would be a placeholder standing in for
/// nothing.
enum RecordStepAction {
    /// Full-width prominent button — `Continue` on `when` / `name`,
    /// `Skip` / `Done` on the optional steps (D3), `Publish` on `privacy`.
    case primary(LocalizedStringResource, enabled: Bool, loading: Bool, () -> Void)
}

/// Shared geometry for every step: a title, a content slot, and one optional
/// button beneath it.
///
/// Steps supply content and a button role and never their own layout, so the
/// title baseline and button position do not drift between screens as the user
/// moves through the flow — which is the whole reason the flow reads as one
/// motion rather than eleven pages.
struct RecordStepScaffold<Content: View>: View {
    let title: LocalizedStringResource
    var subtitle: LocalizedStringResource?
    var action: RecordStepAction?
    /// When true the content fills the space beneath the title instead of
    /// scrolling, and the action floats over its bottom edge. The photo step
    /// uses it so the library grid runs to the bottom of the screen — only
    /// steps whose action is always enabled should, since a disabled pill is
    /// an outline and would not read over a photo.
    var contentFillsHeight: Bool = false
    @ViewBuilder let content: () -> Content

    var body: some View {
        VStack(spacing: Spacing.lg) {
            VStack(spacing: Spacing.xs) {
                Text(title)
                    .pebblesFont(.title)
                    .foregroundStyle(Color.system.foreground)
                    .multilineTextAlignment(.center)

                if let subtitle {
                    Text(subtitle)
                        .pebblesFont(.subhead)
                        .foregroundStyle(Color.system.secondary)
                        .multilineTextAlignment(.center)
                }
            }
            .frame(maxWidth: .infinity)
            .padding(.horizontal, Spacing.lg)

            if contentFillsHeight {
                content()
                    .padding(.horizontal, Spacing.lg)
                    .padding(.bottom, Spacing.sm)
                    .frame(maxWidth: .infinity, maxHeight: .infinity)
                    .overlay(alignment: .bottom) { floatingAction }
            } else {
                ScrollView {
                    content()
                        .padding(.horizontal, Spacing.lg)
                        .padding(.bottom, Spacing.lg)
                }
                .scrollBounceBehavior(.basedOnSize)

                if let action {
                    actionView(action)
                        .padding(.horizontal, Spacing.lg)
                        .padding(.bottom, Spacing.sm)
                }
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
    }

    /// The same pill, laid over the content's bottom edge rather than pushing
    /// it up. The shadow is what separates it from whatever it covers.
    @ViewBuilder
    private var floatingAction: some View {
        if let action {
            actionView(action)
                .shadow(color: .black.opacity(0.22), radius: 14, y: 4)
                .padding(.horizontal, Spacing.xl)
                .padding(.bottom, Spacing.lg)
        }
    }

    @ViewBuilder
    private func actionView(_ action: RecordStepAction) -> some View {
        switch action {
        case let .primary(label, enabled, loading, perform):
            Button(action: perform) {
                Text(label)
            }
            .buttonStyle(PebblesPrimaryButtonStyle(isLoading: loading))
            .disabled(!enabled || loading)
        }
    }
}

File: apps/android/app/src/main/kotlin/app/pbbls/android/features/path/record/RecordStepScaffold.kt
package app.pbbls.android.features.path.record

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextAlign
import app.pbbls.android.components.PebblesPrimaryButton
import app.pbbls.android.theme.PebblesText
import app.pbbls.android.theme.PebblesTheme
import app.pbbls.android.theme.PebblesTypography
import app.pbbls.android.theme.Spacing

/**
 * The single action a step may offer beneath its content — ports iOS
 * `RecordStepAction`.
 */
sealed interface RecordStepAction {
    /** Quiet text button — Skip / Done on the optional steps (M58 D3). */
    data class Text(
        val label: String,
        val onClick: () -> Unit,
    ) : RecordStepAction

    /**
     * Full-width prominent button — Continue on `when` / `name`, Publish on
     * `privacy`.
     */
    data class Primary(
        val label: String,
        val enabled: Boolean,
        val isLoading: Boolean,
        val onClick: () -> Unit,
    ) : RecordStepAction
}

/**
 * Shared geometry for every step: a title, a scrolling content slot, and one
 * optional button beneath it — ports iOS `RecordStepScaffold`.
 *
 * Steps supply content and a button role and never their own layout, so the
 * title baseline and button position do not drift between screens as the user
 * moves through the flow — which is the whole reason the flow reads as one
 * motion rather than eleven pages.
 *
 * Stateless: every input arrives as a parameter, so screenshot previews drive
 * it with no services.
 */
@Composable
fun RecordStepScaffold(
    title: String,
    modifier: Modifier = Modifier,
    subtitle: String? = null,
    action: RecordStepAction? = null,
    /**
     * Whether the content slot scrolls. Pass `false` for content that is already
     * scrollable — see `RecordStep.bringsOwnScroll`; nesting two vertical
     * scrolls throws at measure time.
     */
    contentScrolls: Boolean = true,
    content: @Composable () -> Unit,
) {
    val system = PebblesTheme.colors.system
    Column(
        modifier = modifier.fillMaxSize(),
        verticalArrangement = Arrangement.spacedBy(Spacing.lg),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        Column(
            modifier = Modifier.fillMaxWidth().padding(horizontal = Spacing.lg),
            verticalArrangement = Arrangement.spacedBy(Spacing.xs),
            horizontalAlignment = Alignment.CenterHorizontally,
        ) {
            PebblesText(
                text = title,
                style = PebblesTypography.title,
                color = system.foreground,
                textAlign = TextAlign.Center,
            )
            if (subtitle != null) {
                PebblesText(
                    text = subtitle,
                    style = PebblesTypography.subhead,
                    color = system.secondary,
                    textAlign = TextAlign.Center,
                )
            }
        }

        val scrollState = rememberScrollState()
        Column(
            modifier =
                Modifier
                    .fillMaxWidth()
                    .weight(1f)
                    .then(if (contentScrolls) Modifier.verticalScroll(scrollState) else Modifier)
                    .padding(horizontal = Spacing.lg)
                    .padding(bottom = Spacing.lg),
            horizontalAlignment = Alignment.CenterHorizontally,
        ) {
            content()
        }

        when (action) {
            null -> Unit
            is RecordStepAction.Text ->
                TextButton(
                    onClick = action.onClick,
                    modifier = Modifier.padding(bottom = Spacing.sm),
                ) {
                    PebblesText(
                        text = action.label,
                        style = PebblesTypography.callout,
                        color = system.secondary,
                    )
                }
            is RecordStepAction.Primary ->
                PebblesPrimaryButton(
                    text = action.label,
                    onClick = action.onClick,
                    enabled = action.enabled && !action.isLoading,
                    isLoading = action.isLoading,
                    modifier =
                        Modifier
                            .fillMaxWidth()
                            .padding(horizontal = Spacing.lg)
                            .padding(bottom = Spacing.sm),
                )
        }
    }
}
